txnsync/txngroups.go

```diff
@@ -17,23 +17,417 @@
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Receiver        []byte              `codec:"rcv,allocbound=maxAddressBytes"`
+	BitmaskReceiver bitmask             `codec:"rcvbm"`
+	Amount          []basics.MicroAlgos `codec:"amt,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAmount   bitmask             `codec:"amtbm"`
+
+	CloseRemainderTo        []byte  `codec:"close,allocbound=maxAddressBytes"`
+	BitmaskCloseRemainderTo bitmask `codec:"closebm"`
+}
+
+type encodedAssetConfigTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	ConfigAsset        []basics.AssetIndex `codec:"caid,allocbound=maxEncodedTransactionGroup"`
+	BitmaskConfigAsset bitmask             `codec:"caidbm"`
+
+	AssetParams        []basics.AssetParams `codec:"apar,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAssetParams bitmask              `codec:"aparbm"`
+}
+
+type encodedAssetTransferTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	XferAsset        []basics.AssetIndex `codec:"xaid,allocbound=maxEncodedTransactionGroup"`
+	BitmaskXferAsset bitmask             `codec:"xaidbm"`
+
+	AssetAmount        []uint64 `codec:"aamt,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAssetAmount bitmask  `codec:"aamtbm"`
+
+	AssetSender        []byte  `codec:"asnd,allocbound=maxAddressBytes"`
+	BitmaskAssetSender bitmask `codec:"asndbm"`
+
+	AssetReceiver        []byte  `codec:"arcv,allocbound=maxAddressBytes"`
+	BitmaskAssetReceiver bitmask `codec:"arcvbm"`
+
+	AssetCloseTo        []byte  `codec:"aclose,allocbound=maxAddressBytes"`
+	BitmaskAssetCloseTo bitmask `codec:"aclosebm"`
+}
+
+type encodedAssetFreezeTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	FreezeAccount        []byte  `codec:"fadd,allocbound=maxAddressBytes"`
+	BitmaskFreezeAccount bitmask `codec:"faddbm"`
+
+	FreezeAsset        []basics.AssetIndex `codec:"faid,allocbound=maxEncodedTransactionGroup"`
+	BitmaskFreezeAsset bitmask             `codec:"faidbm"`
+
+	AssetFrozen        []bool  `codec:"afrz,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAssetFrozen bitmask `codec:"afrzbm"`
+}
+
+//msgp:allocbound applicationArgs transactions.EncodedMaxApplicationArgs
+type applicationArgs [][]byte
+
+//msgp:allocbound addresses transactions.EncodedMaxAccounts
+type addresses []basics.Address
+
+//msgp:allocbound appIndeces transactions.EncodedMaxForeignApps
+type appIndeces []basics.AppIndex
+
+//msgp:allocbound assetIndeces transactions.EncodedMaxForeignAssets
+type assetIndeces []basics.AssetIndex
+
+//msgp:allocbound program config.MaxAppProgramLen
+type program []byte
+
+type encodedApplicationCallTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	ApplicationID        []basics.AppIndex `codec:"apid,allocbound=maxEncodedTransactionGroup"`
+	BitmaskApplicationID bitmask           `codec:"apidbm"`
+
+	OnCompletion        []transactions.OnCompletion `codec:"apan,allocbound=maxEncodedTransactionGroup"`
```

Comment:
this field could really benefit from better packing. `transactions.OnCompletion` contains only few values, and we could easily fit 2 of them per a single byte.
Applications are going to be pretty common - so we'd better optimize the packer for that.